component/azstorage/file_share.go

```diff
@@ -550,29 +745,150 @@ func (fs *FileShare) WriteFromBuffer(name string, metadata map[string]string, da
+	}
+
+	// This is not currently supported for a fileshare account
 	return syscall.ENOTSUP
 }
-func (fs *FileShare) ChangeOwner(string, int, int) error {
-	return syscall.ENOTSUP
+
+// StageAndCommit : write data to an Azure file given a list of ranges
+func (fs *FileShare) StageAndCommit(name string, bol *common.BlockOffsetList) error {
+	log.Trace("FileShare::StageAndCommit : name %s", name)
+
+	fileName, dirPath := getFileAndDirFromPath(filepath.Join(fs.Config.prefixPath, name))
+	fileURL := fs.Share.NewDirectoryURL(dirPath).NewFileURL(fileName)
+
+	var data []byte
+
+	for _, rng := range bol.BlockList {
+		if rng.Truncated() {
+			data = make([]byte, rng.EndIndex-rng.StartIndex)
+			rng.Flags.Clear(common.TruncatedBlock)
+		} else {
+			data = rng.Data
+		}
+		if rng.Dirty() {
+			_, err := fileURL.UploadRange(context.Background(),
+				rng.StartIndex,
+				bytes.NewReader(data),
+				nil,
+			)
+			if err != nil {
+				log.Err("FileShare::StageAndCommit : Failed to upload range to file %s at index %v (%s)", name, rng.StartIndex, err.Error())
+				return err
+			}
+			rng.Flags.Clear(common.DirtyBlock)
+		}
+	}
+	return nil
 }
-func (fs *FileShare) TruncateFile(string, int64) error {
-	return syscall.ENOTSUP
+
+// Write : write data at given offset to an Azure file
+func (fs *FileShare) Write(options internal.WriteFileOptions) (err error) {
+	name := options.Handle.Path
+	offset := options.Offset
+	data := options.Data
+	// length := int64(len(options.Data))
+
+	defer log.TimeTrack(time.Now(), "FileShare::Write", options.Handle.Path)
+	log.Trace("FileShare::Write : name %s offset %v", name, offset)
+
+	if len(data) == 0 {
+		return nil
+	}
+
+	fileName, dirPath := getFileAndDirFromPath(filepath.Join(fs.Config.prefixPath, name))
+	fileURL := fs.Share.NewDirectoryURL(dirPath).NewFileURL(fileName)
+
+	// fileOffsets, err := fs.GetFileBlockOffsets(name)
+	// if err != nil {
+	// 	log.Err("FileShare::Write : Failed to get file range offsets %s", err.Error())
+	// 	return err
+	// }
+
+	// _, _, exceedsFileBlocks, _ := fileOffsets.FindBlocksToModify(offset, length) // **********but this method only looks for true file size rather than file capacity
+	// if exceedsFileBlocks {
+	// 	err = fs.TruncateFile(name, offset+length)
+	// 	if err != nil {
+	// 		log.Err("FileShare::Write : Failed to truncate Azure file %s", err.Error())
+	// 		return err
+	// 	}
+	// }
+
+	_, err = fileURL.UploadRange(context.Background(), options.Offset, bytes.NewReader(data), nil)
+	if err != nil {
+		log.Err("FileShare::Write : Failed to write data to Azure file %s", err.Error())
+		return err
+	}
+
+	return nil
 }
-func (fs *FileShare) StageAndCommit(name string, bol *common.BlockOffsetList) error {
-	return syscall.ENOTSUP
+
+// GetFileBlockOffsets : store file range list and corresponding offsets
+func (fs *FileShare) GetFileBlockOffsets(name string) (shareFileRangeList *common.BlockOffsetList, err error) {
+	log.Trace("FileShare::GetFileBlockOffsets : name %s", name)
```

Comment:
Do we need these methods here ? It shall only be required if we wish to support file-share over streaming.

Reply:
Should be fine to implement. There's no harm in supporting it is there?

Reply:
There is no harm as such but if read/write streaming is not possible for files then we can just have dummy implementation.

Reply:
I think writes streaming is actually more suited to file shares since they have very convenient Upload APIs

Reply:
what was the conclusion of the conversation regarding file creating and pre-specifying the size of the file? the setting we set on the top just dictates the max size of the file, but the file can be of size 0 and we can keep adding to it, correct?

Reply:
we're going to have to resize when we append basically.